doc-site/docs/tutorials/custom_contracts/cardano.md

```diff
@@ -0,0 +1,542 @@
+    // and those UTxOs contain some amount of ADA and native tokens.
+    // You can't spent part of a UTxO in a transaction; instead, transactions
+    // include inputs with more funds than they need, and a "change" output
+    // to give any excess funds back to the original sender.
+
+    // Build a transaction with
+    //  - One or more inputs containing at least `amount` ADA at the address `from_address`
+    //  - One output containing exactly `amount` ADA at the address `to_address`
+    //  - One output containing any change at the address `from_address`
+    let tx = TxBuilder::new()
+        .with_input(CoinSelectionInput(address_source.clone(), req.amount))
+        .with_output(
+            OutputBuilder::new()
+                .address(req.to_address.clone())
+                .with_value(req.amount),
+        )
+        .with_output(FeeChangeReturn(address_source));
+
+    // Return that TX. The framework will sign, submit, and monitor it.
+    // By returning a `NewMonitoredTx`, we tell the framework that we want it to monitor this transaction.
+    // This enables the TransactionApproved, TransactionRolledBack, and TransactionFinalized events from before.
+    // Note that we decide the transaction has been finalized after 4 blocks have reached the chain.
+    Ok(NewMonitoredTx(
+        Box::new(tx),
+        FinalizationCondition::AfterBlocks(4),
+    ))
+}
+
+/// This function is called when a TX produced by this contract is submitted to the blockchain, but before it has reached a block.
+fn handle_submit(_: Config<()>, tx: SubmittedTx) -> WorkerResult<Ack> {
+    // Keep track of which TXs have been submitted.
+    let mut state: CurrentState = kv::get("current_state")?.unwrap_or_default();
+    state.submitted_txs.insert(tx.hash);
+    kv::set("current_state", &state)?;
+
+    Ok(Ack)
+}
+
+fn query_current_state(_: Config<()>, _: Params<()>) -> WorkerResult<Json<CurrentState>> {
+    Ok(Json(kv::get("current_state")?.unwrap_or_default()))
+}
+
+#[balius_sdk::main]
+fn main() -> Worker {
+    Worker::new()
+        .with_request_handler("send_ada", FnHandler::from(send_ada))
+        .with_request_handler("current_state", FnHandler::from(query_current_state))
+        .with_tx_submitted_handler(handle_submit)
+}
+
+```
+
+## Deploying the dApp
+
+You can use the `firefly-cardano-deploy` tool to deploy this dApp to your running FireFly instance.
+This tool will
+ - Compile your dApp to WebAssembly
+ - Deploy that WebAssembly to a running FireFly node 
+ - Deploy your interface to that FireFly node
+ - Create an API for that interface
+
+```sh
+# The version here should match the version of firefly-cardano which you are using.
+cargo install --git https://github.com/hyperledger/firefly-cardano --version 0.3.1 firefly-cardano-deploy
+
+CONTRACT_PATH="/path/to/your/dapp"
+FIREFLY_URL="http://localhost:5000"
+firefly-cardano-deploy --contract-path "$CONTRACT_PATH" --firefly-url "$FIREFLY_URL"
```

Comment:
This is great, probably as a follow up adding this to the FireFly CLI would be great :) We have similar deploy and install commands in for Fabric and EVM

Reply:
That would definitely be more convenient than asking users to install some standalone tool. But part of deploying these contracts is compiling rust code to webassembly. Would it be OK for the FireFly CLI to shell out to `cargo`?

Reply:
oh that is good point, as all these tools are in Rust! I think it would be fine